src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetLatestDiagnosticsForSpan.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV1
+{
+    internal partial class DiagnosticIncrementalAnalyzer : BaseDiagnosticIncrementalAnalyzer
+    {
+        private class LatestDiagnosticsForSpanGetter
+        {
+            private readonly DiagnosticIncrementalAnalyzer _owner;
+
+            private readonly Document _document;
+            private readonly TextSpan _range;
+            private readonly bool _blockForData;
+            private readonly CancellationToken _cancellationToken;
+
+            private readonly DiagnosticAnalyzerDriver _spanBasedDriver;
+            private readonly DiagnosticAnalyzerDriver _documentBasedDriver;
+            private readonly DiagnosticAnalyzerDriver _projectDriver;
+
+            private readonly List<DiagnosticData> _diagnostics;
+
+            public LatestDiagnosticsForSpanGetter(
+                DiagnosticIncrementalAnalyzer owner, Document document, SyntaxNode root, TextSpan range, bool blockForData, CancellationToken cancellationToken) :
+                this(owner, document, root, range, blockForData, new List<DiagnosticData>(), cancellationToken)
+            {
+            }
+
+            public LatestDiagnosticsForSpanGetter(
+                DiagnosticIncrementalAnalyzer owner, Document document, SyntaxNode root, TextSpan range, bool blockForData, List<DiagnosticData> diagnostics, CancellationToken cancellationToken)
+            {
+                _owner = owner;
+
+                _document = document;
+                _range = range;
+                _blockForData = blockForData;
+                _cancellationToken = cancellationToken;
+
+                _diagnostics = diagnostics;
+
+                // Share the diagnostic analyzer driver across all analyzers.
+                var fullSpan = root == null ? null : (TextSpan?)root.FullSpan;
+
+                _spanBasedDriver = new DiagnosticAnalyzerDriver(_document, _range, root, _owner._diagnosticLogAggregator, _owner.HostDiagnosticUpdateSource, _cancellationToken);
+                _documentBasedDriver = new DiagnosticAnalyzerDriver(_document, fullSpan, root, _owner._diagnosticLogAggregator, _owner.HostDiagnosticUpdateSource, _cancellationToken);
+                _projectDriver = new DiagnosticAnalyzerDriver(_document.Project, _owner._diagnosticLogAggregator, _owner.HostDiagnosticUpdateSource, _cancellationToken);
+            }
+
+            public List<DiagnosticData> Diagnostics => _diagnostics;
```

Comment:
The property could be merged with its backing field into a single read-only property `public List<DiagnosticData> Diagnostics { get; }`.